crates/services/sync/src/import/cache.rs

```diff
@@ -0,0 +1,541 @@
+            (CachedDataBatch::Blocks(blocks_batch), CachedData::Header(header)) => {
+                debug_assert_eq!(blocks_batch.range.end, height);
+                Self::push_current_chunk(
+                    chunks,
+                    CachedDataBatch::Blocks(blocks_batch),
+                    max_chunk_size,
+                );
+                CachedDataBatch::Headers(Batch::new(
+                    None,
+                    height..height.saturating_add(1),
+                    vec![header],
+                ))
+            }
+        }
+    }
+
+    fn push_missing_chunks(
+        chunks: &mut Vec<CachedDataBatch>,
+        current_height: u32,
+        height: u32,
+        max_chunk_size: usize,
+        chunk_size_u32: u32,
+        end: u32,
+    ) {
+        let missing_chunks =
+            (current_height..height)
+                .step_by(max_chunk_size)
+                .map(move |chunk_start| {
+                    let block_end = chunk_start.saturating_add(chunk_size_u32).min(end);
+                    CachedDataBatch::None(chunk_start..block_end)
+                });
+        chunks.extend(missing_chunks);
+    }
+
+    pub fn remove_element(&mut self, height: &u32) {
+        let mut lock = self.0.lock();
+        lock.remove(height);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::import::{
+        cache::{
+            Cache,
+            CachedData,
+            CachedDataBatch,
+        },
+        Batch,
+    };
+    use fuel_core_types::{
+        blockchain::{
+            block::Block,
+            consensus::Sealed,
+            header::BlockHeader,
+        },
+        fuel_tx::Bytes32,
+        tai64::Tai64,
+    };
+    use futures::StreamExt;
+    use std::ops::{
+        Range,
+        RangeInclusive,
+    };
+    use test_case::test_case;
+
+    fn create_header(height: u32) -> Sealed<BlockHeader> {
+        Sealed {
+            entity: BlockHeader::new_block(height.into(), Tai64::from_unix(0)),
+            consensus: Default::default(),
+        }
+    }
+    fn create_block(height: u32) -> Sealed<Block> {
+        Sealed {
+            entity: Block::new(
+                (&create_header(height).entity).into(),
+                Vec::new(),
+                &[],
+                Bytes32::default(),
+            )
+            .unwrap(),
+            consensus: Default::default(),
+        }
+    }
+
+    #[test_case(&[], &[], 3, 0..=10 => vec![
+        CachedDataBatch::None(0..3),
+        CachedDataBatch::None(3..6),
+        CachedDataBatch::None(6..9),
+        CachedDataBatch::None(9..11),
+    ] ; "multiple empty chunks")]
+    #[test_case(&[
+        create_header(0)
+    ], &[], 3, 0..=10 => vec![
+        CachedDataBatch::Headers(Batch::new(None, 0..1, vec![create_header(0)])),
+        CachedDataBatch::None(1..4),
+        CachedDataBatch::None(4..7),
+        CachedDataBatch::None(7..10),
+        CachedDataBatch::None(10..11),
+    ]; "one header and empty ranges")]
```

Comment:
The max batch 0 created a panic and it makes no sens so I changed to `NonZeroUsize` however I didn't propagate up to the config to not cause too much changes I opened an issue for that :  https://github.com/FuelLabs/fuel-core/issues/2462
For the empty range it's not really possible to do it here because the type on test is `RangeInclusive` which force me to have 1 element at least.

Reply:
Ah right, nice :+1: